src/config.rs

```diff
@@ -0,0 +1,376 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use miette::{bail, ensure, miette, Context, IntoDiagnostic};
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+use crate::{
+    manifest::{Edition, CANARY_EDITION},
+    registry::{RegistryRef, RegistryUri},
+};
+
+/// Location of the configuration file
+const CONFIG_FILE: &str = ".buffrs/config.toml";
+
+/// Key for common default arguments
+const DEFAULT_ARGS_KEY: &str = "*";
+
+/// Representation of the configuration file
+///
+/// # Example
+///
+/// ```toml
+/// edition = "0.10.0"
+///
+/// [registries]
+/// some_org = "https://artifactory.example.com/artifactory/some-org"
+///
+/// [registry]
+/// default = "some_org"
+///
+/// [commands]
+/// default_args = ["--insecure"]
+///
+/// [commands.install]
+/// default_args = ["--generate-buf-yaml"]
+/// ```
+///
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Config {
+    /// Edition of this configuration file (in sync with the Proto.toml edition)
+    edition: Edition,
+
+    /// Path to the configuration file
+    config_path: Option<PathBuf>,
+
+    /// Default registry alias to use if none is specified
+    default_registry: Option<String>,
+
+    /// List of registries
+    registries: HashMap<String, RegistryUri>,
+
+    /// Default arguments for commands
+    command_defaults: HashMap<String, Vec<String>>,
+}
+
+impl Config {
+    /// Create a new configuration with default values
+    /// # Arguments
+    /// * `cwd` - Starting directory to search for the configuration file
+    ///
+    pub fn new(cwd: Option<&Path>) -> miette::Result<Self> {
+        match Self::locate_config(cwd) {
+            Some(config_path) => Self::new_from_config_file(&config_path),
+            None => Ok(Self {
+                edition: Edition::latest(),
+                config_path: None,
+                default_registry: None,
+                registries: HashMap::new(),
+                command_defaults: HashMap::new(),
+            }),
+        }
+    }
+
+    /// Parse a registry argument
+    ///
+    /// # Arguments
+    /// * `registry` - The registry argument to parse
+    ///
+    /// # Returns
+    /// URI with either alias scheme or actual URI:
+    /// - <alias> -> alias://<alias>
+    /// - <uri> -> <uri>
+    /// - None -> alias://<default>
+    pub fn parse_registry_arg(&self, registry: &Option<String>) -> miette::Result<RegistryRef> {
+        match registry {
+            Some(registry) => RegistryRef::from_str(registry),
```

Comment:
Nit: .parse is more idiomatic imo

Reply:
0f35c69d2b211fca144ae8cbcec13344b9a4200e